File: 0039/main.py
from collections import deque
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = list()
        q = deque()
        
        for c in candidates:
            if c == target:
                result.append([c])
            else:
                q.append((target - c, [c]))
                
        while len(q) > 0:
            val, combo = q.pop()
            
            for c in candidates:
                if c < combo[-1]: 
                    continue
                
                tmp = val - c
                if tmp == 0:
                    result.append(sorted(combo[:] + [c]))
                elif tmp > 0:
                    q.append((tmp, combo[:] + [c]))
            
        return result

if __name__ == "__main__":
    res = Solution().combinationSum([2, 3, 6, 7], 7)
    assert([7] in res and [2, 2, 3] in res)

File: 0039/main.rs
struct Solution { }

use std::collections::VecDeque;

impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut q: VecDeque<(i32, Vec<i32>)> = VecDeque::new();
        
        for c in candidates.iter() {
            if c == &target {
                result.push(vec![*c]);
            } else {
                q.push_back((target - c, vec![*c]));
            }
        }
        
        while q.len() > 0 {
            let (value, combo) = q.pop_front().unwrap();
            
            let last: &i32 = combo.last().unwrap();
            for c in candidates.iter() {
                if c < last {
                    continue;
                }
                
                let tmp = value - c;
                if tmp == 0 {
                    let mut res = combo.clone();
                    res.push(*c);
                    result.push(res);
                } else if tmp > 0 {
                    let mut res = combo.clone();
                    res.push(*c);
                    q.push_back((tmp, res));
                }
            }
        }
        
        return result;
    }
}

fn main() {
    let result = Solution::combination_sum(vec![2, 3, 6, 7], 7);
    assert!(result.contains(&vec![7i32]));
    assert!(result.contains(&vec![2i32, 2, 3]));
}